runtime/gc_vlhgc/VLHGCAccessBarrier.cpp

```diff
@@ -249,17 +250,24 @@ MM_VLHGCAccessBarrier::postStoreClassToClassLoader(J9VMThread *vmThread, J9Class
 }
 
 IDATA
-MM_VLHGCAccessBarrier::indexableDataDisplacement(J9VMThread *vmThread, J9IndexableObject *src, J9IndexableObject *dst)
+MM_VLHGCAccessBarrier::indexableDataDisplacement(J9StackWalkState *walkState, J9IndexableObject *src, J9IndexableObject *dst)
 {
 	IDATA displacement = 0;
 
 #if defined(J9VM_ENV_DATA64)
 	Assert_MM_true(_extensions->isVirtualLargeObjectHeapEnabled);
-	/* Adjacency check against dst object since src object may be overwritten during sliding compaction. */
-	if (_extensions->indexableObjectModel.isDataAdjacentToHeader(dst))
+	/* Potential danger this is not called by a GC RootScanner */
+	Assert_MM_true(walkState->objectSlotWalkFunction == gc_vmThreadStackDoOSlotIterator);
```

Comment:
@dmitripivkine This might be a bit overkill assert, for which I had to extern the iterator callback. I can remove it (and undo externing), if you prefer.

Before, JIT would do displacement math for any scanner (even non GC ones), in which case we would not find RootScanner in userData2 of WalkState. But I enhanced JIT to call displacement math only if object moved, which is I guess safe to assume that can happen only by GC (and only specific GC scanners).